File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function (lists) {
    var mergeTwoLists = function (list1, list2) {
   
        if (!list1) {
            return list2
        }
        if (!list2) {
            return list1
        }
        if (list1.val < list2.val) {
            list1.next = mergeTwoLists(list1.next, list2);
            return list1

        }
        else {
            list2.next = mergeTwoLists(list1, list2.next);
            return list2
        }
    };
    function solve(lists) {
        if (lists.length==2) {
            return  mergeTwoLists(lists[0], lists[1])

        }
        return  mergeTwoLists(lists[0], solve(lists.slice(1)))
    }
    const n = lists.length;
    if (n == 0) {
        return null;
    }
    if (n==1){
        return lists[0]
    }

    const res = solve(lists)

    return res;


};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLinkedList(ListNode* list1, ListNode* list2) {
        if (list1 == NULL) {
            return list2;
        }
        if (list2 == NULL) {
            return list1;
        }
        if (list1->val < list2->val) {
            list1->next = mergeTwoLinkedList(list1->next, list2);
            return list1;
        } else {
            list2->next = mergeTwoLinkedList(list2->next, list1);
            return list2;
        }
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0) {
            return NULL;
        }
        // my base case
        if (lists.size() == 1) {

            return lists[0];
        }
        vector <ListNode*> right(lists.begin() + 1, lists.end());

        return mergeTwoLinkedList(lists[0], mergeKLists(right));
    }
};